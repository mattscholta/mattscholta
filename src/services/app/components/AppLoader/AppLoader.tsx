import * as React from 'react';
import classnames from 'classnames';
import { useRecoilValue } from 'recoil';

import { appAtom } from '@src/recoil/atoms/app';

import styles from './styles.module.css';

interface AppLoaderProps {
  className?: string;
  message?: string;
  state?: boolean;
}

/**
 * @name AppLoader
 * @description A generic loading component that uses a CSS box that
 * we spin using a CSS animation
 */
const AppLoader: React.FC<AppLoaderProps> = (props) => {
  const { className, message = 'Loading', state } = props;

  // Hooks
  const { loading } = useRecoilValue(appAtom);

  // Styles
  const tailwind = `u-flex u-items-center u-justify-center u-fixed u-inset-0`;
  const cssComponent = classnames(tailwind, className, styles.component, {
    [styles.visible]: state || loading
  });

  return (
    <div className={cssComponent}>
      <div className="u-flex u-flex-col u-items-center">
        <div className={styles.loader} />
        <p>{message}</p>
      </div>
    </div>
  );
};

export { AppLoader };
export type { AppLoaderProps };
