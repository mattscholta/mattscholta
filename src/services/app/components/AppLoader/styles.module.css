.component {
  background-color: rgba(255 255 255 / 85%);
  opacity: 0;
  pointer-events: none;
  transition: opacity 200ms ease-out;
  z-index: 8;
}

.component p {
  color: theme('colors.gray.4');
  font-size: theme('fontSize.18');
  margin-top: theme('spacing.2x');
}

.loader {
  animation: 'loader' 2000ms infinite linear;
  border: theme('spacing.1xs') solid theme('colors.gray.2');
  border-left-color: theme('colors.primary');
  border-radius: 50%;
  height: 50px;
  width: 50px;
}

.visible {
  opacity: 1;
  pointer-events: all;
}
