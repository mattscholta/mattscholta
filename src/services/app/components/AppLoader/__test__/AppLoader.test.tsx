import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { RecoilRoot } from 'recoil';

import { AppLoader, AppLoaderProps } from '../AppLoader';

describe('AppLoader', () => {
  let props: AppLoaderProps;
  let wrapper: RenderResult;

  const createWrapper = () => (
    <RecoilRoot>
      <AppLoader {...props} />
    </RecoilRoot>
  );

  beforeEach(() => {
    props = {};
    wrapper = render(createWrapper());
  });

  it('matches snapshot', () => {
    expect(wrapper.container.childNodes).toMatchSnapshot();
  });
});
