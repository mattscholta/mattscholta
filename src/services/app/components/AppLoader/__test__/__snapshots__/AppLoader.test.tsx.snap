// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLoader matches snapshot 1`] = `
NodeList [
  <div
    class="u-flex u-items-center u-justify-center u-fixed u-inset-0 component"
  >
    <div
      class="u-flex u-flex-col u-items-center"
    >
      <div
        class="loader"
      />
      <p>
        Loading
      </p>
    </div>
  </div>,
]
`;
